Initial Django project and API app

willbarton opened the pull request:
Here's an initial Django project and API app. The API app has models for repositories and data points, tests for the models, mock-data sources for generating some random data points to test with, and a basic REST API for both. 

@Ooblioob @mehtadev17, can you review for anything that stands out? It's a lot, but a lot of it is Django boilerplate (pretty much all of porchlight/).

willbarton committed:
Basic Porchlight Django project and API app

Basic Django project in ‘porchlight’, plus a ‘porchlightapi’ app with
the initial data models, tests, and some mock data sources. No Rest API
yet.

willbarton committed:
First attempt at a REST API

willbarton committed:
Added missing 'created' datetime to data points.

willbarton committed:
Add filtering to repository name, project in API

willbarton committed:
Use DRF's built-in search filter

willbarton committed:
Add search for DataPoints by fields on their related repository

willbarton committed:
Add ordering filter

Ooblioob commented on porchlightapi/models.py:
```diff
@@ -0,0 +1,161 @@
+
+class Repository(models.Model):
+    """
+    A repository of code that is deployed that may or may not have
```
:+1: for documentation!

willbarton committed:
Realize that some sources, like the mock source, need not be specific to deployed or underplayed

Ooblioob commented:
@willbarton 
I see what you're trying to do here with the `reliable_mock_undeployed_source` method but this may be a place where a formal Mocking framework is more useful.  What we want is a method we can mock in order to force it to return a set value.  We could leave `mock_undeployed_source` (perhaps we rename it) since it generates visually pretty data, but for our tests we should just call the test like this: 

```
@patch("porchlightapi.sources.mock_undeployed_source")
def test_undeployed_value_source(self, mock_undeployed_source):
    mock_undeployed_source.return_value = (value1, value2, value3)
...etc
```

This will give us control over the method even if it's stored in the db (I think, I've never tried it with a method like this).  We can control what it returns and even verify that it was called (when appropriate).  

If we use a mocking class then we don't have to write our own extra stuff and it's a little more standard.  The code you've written is fine for now, but because the "mock" implementation is random it's a nightmare to test and requires us to write multiple "reliable mock" implementations in our unit tests!

Here's the link to the documentation on Mocks  
http://www.voidspace.org.uk/python/mock/getting-started.html#using-mock

willbarton committed:
Revert "Realize that some sources, like the mock source, need not be specific to deployed or underplayed"

This reverts commit f167ec849b150a5e533e270c2c5a37fa6f79040d.

willbarton commented:
@Ooblioob, Sounds good, I'll pull in mock.

Ooblioob commented on porchlightapi/tests.py:
```diff
@@ -0,0 +1,105 @@
+    def setUp(self):
+        create_test_repository()
+
+    def test_value_data_point_creation(self):
```
I see no strong reason we need to make an integration test out of this.  We could accomplish the same by mocking the test_repo object.  Something to the tune of:

```
@patch("porchlightapi.models.Repository")
def test_value_data_point_creation(self, test_repo):
    test_repo.undeployed_value.return_value = some_value
    ...
```

There's also a way to mock `self.create` so we can just call the `assert_called_with()` method to verify it was called with what we wanted.  
http://www.voidspace.org.uk/python/mock/getting-started.html#using-mock

willbarton committed:
Switch to using mock for mocking value functions.

willbarton committed:
Add a test for ValueDataPointManager's create_datapoint method

willbarton committed:
Remove possible confusion relating to word "mock"

Remove possible confusion between my own "mock" sources, which generate
random source data, and the mock library.

willbarton commented:
@Ooblioob, when you get a chance, can you review the latest commits? I've switched to using mock (which also makes the test code cleaner, which makes me happy), and I've clarified the `ValueDataPoint` test case by making it a test case for `ValueDataPointManager`, testing that it correctly extracts data from a `Repository` (mocked) and correctly calls the default manager `create()` method (which I'm confirming using mock).

willbarton committed:
Add some documentation to the Rest API

The API viewset docstring is displayed by Django Rest Framework in its
browsable API.

willbarton committed:
Add pagination to data points API and document it.

Ooblioob commented:
@willbarton Very nice job, you get a selfie!  The changes for the mocks look great!  
![selfie-3](http://i.imgur.com/rkGm95w.gif)

Prepare to merge!
